msal/src/main/java/com/microsoft/identity/client/controllers/LocalMSALController.java

```diff
@@ -134,26 +140,138 @@ private TokenResult performTokenRequest(OAuth2Strategy strategy, AuthorizationRe
+                acquireTokenSilentResult.setAuthenticationResult(authenticationResult);
+            } else {
+                // TODO Is this the correct Exception?
+                throw new MsalClientException(MsalUiRequiredException.NO_TOKENS_FOUND, "No refresh token was found. ");
+            }
+        } else if (!scopesMatch(parameters.getScopes(), cacheRecord.getAccessToken().getTarget())) {
+            // TODO Do we want to indicate that no MATCHING RT was found?
+            throw new MsalClientException(MsalUiRequiredException.NO_TOKENS_FOUND, "No refresh token was found. ");
+        } else if (cacheRecord.getAccessToken().isExpired()) {
+            tokenCache.removeCredential(cacheRecord.getAccessToken());
+
+            parameters.setRefreshToken(cacheRecord.getRefreshToken());
+
+            final TokenResult tokenResult = performSilentTokenRequest(strategy, parameters);
+            acquireTokenSilentResult.setTokenResult(tokenResult);
+
+            // TODO Do I need to check the success state of the TokenResult?
+            final ICacheRecord savedRecord = tokenCache.save(
+                    strategy,
+                    getAuthorizationRequest(strategy, parameters),
+                    tokenResult.getTokenResponse()
+            );
+
+            final AuthenticationResult authenticationResult = new AuthenticationResult(savedRecord);
+
+            acquireTokenSilentResult.setAuthenticationResult(authenticationResult);
+        } else {
+            // the result checks out, return that....
+            acquireTokenSilentResult.setAuthenticationResult(
+                    new AuthenticationResult(cacheRecord)
+            );
+        }
+
+        return acquireTokenSilentResult;
+    }
+
+    private boolean refreshTokenIsNull(ICacheRecord cacheRecord) {
+        return null == cacheRecord.getRefreshToken();
+    }
+
+    private boolean refreshTokenCanBeUsed(MSALAcquireTokenSilentOperationParameters parameters, ICacheRecord cacheRecord) {
+        return null != cacheRecord.getRefreshToken() && scopesMatch(parameters.getScopes(), cacheRecord.getRefreshToken().getTarget());
+    }
+
+    private boolean accessTokenIsNull(ICacheRecord cacheRecord) {
+        return null == cacheRecord.getAccessToken();
+    }
+
+    private boolean scopesMatch(final List<String> requestParameters,
+                                final String scopesStr) {
+        final String[] scopes = scopesStr.split("\\s");
+        final List<String> atScopes = Arrays.asList(scopes);
+        return atScopes.containsAll(requestParameters);
+    }
+
+    private TokenResult performSilentTokenRequest(final OAuth2Strategy strategy,
+                                                  final MSALAcquireTokenSilentOperationParameters parameters) throws MsalClientException, IOException {
+        throwIfNetworkNotAvailable(parameters.getAppContext());
+
+        final TokenRequest request = new MicrosoftStsTokenRequest();
```

Comment:
Concrete type.  We likely should add another createTokenRequest method on the strategy.  There's a design problem here to work on .... relative to the type of token request... but I think we can defer that for now, by adding a createRefreshTokenRequest abstract method to oauth2strategy